### PR TITLE
perf(django): patch connection on registration

### PR DESCRIPTION
**Pull request by P403n1x87**
## Description

Patch a Django connection only when it is registered with the `ConnectionHandler` rather than when the `all` iter is invoked, thus avoiding unnecessary work when connections are, e.g., closed.

## Checklist
- [ ] Added to the correct milestone.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] ~~Library documentation is updated.~~
- [ ] ~~[Corp site](https://github.com/DataDog/documentation/) documentation is updated (link to the PR).~~

### CODE REVIEW

**Comment by majorgreys** [comment]
LGTM just needs release notes!

**Review by brettlangdon — COMMENTED** [review]
is it possible for this to miss a connection?

what happens if a db connection is created before our patching code is called?

**Comment by P403n1x87** [comment]
> is it possible for this to miss a connection?
> 
> what happens if a db connection is created before our patching code is called?

Looking at https://github.com/django/django/blob/f55f3ce831fa885dfef0b222c254bb4bf3ca99ef/django/utils/connection.py#L34, it shouldn't be possible to miss any connections by wrapping `__setitem__`. Any new connections that get registered with the `ConnectionHandler` seem to be routed through `__setitem__`. And registration seems to happen soon after a connection is created: https://github.com/django/django/blob/f55f3ce831fa885dfef0b222c254bb4bf3ca99ef/django/utils/connection.py#L62-L63

**Comment by brettlangdon** [comment]
> > is it possible for this to miss a connection?
> > what happens if a db connection is created before our patching code is called?
> 
> Looking at https://github.com/django/django/blob/f55f3ce831fa885dfef0b222c254bb4bf3ca99ef/django/utils/connection.py#L34, it shouldn't be possible to miss any connections by wrapping `__setitem__`. Any new connections that get registered with the `ConnectionHandler` seem to be routed through `__setitem__`. And registration seems to happen soon after a connection is created: https://github.com/django/django/blob/f55f3ce831fa885dfef0b222c254bb4bf3ca99ef/django/utils/connection.py#L62-L63

When are connections created? is it possible to have connections created in `settings.py` / e.g. is there a chance a connection can be created before our patching code is called?

I am thinking less of people using `ddtrace-run`, and more if someone does manual late patching.

This might not be a big deal, since late patching generally causes issues, **but** if it is easy to support late patching probably will mean it is more resilient (?)

**Comment by P403n1x87** [comment]
> I am thinking less of people using `ddtrace-run`, and more if someone does manual late patching.
> 
> This might not be a big deal, since late patching generally causes issues, **but** if it is easy to support late patching probably will mean it is more resilient (?)

All the `settings.py` that I have seen do not import `django` so shouldn't open any connections to any DB at that point. Of course, if `patch` is called once the application has been started, there is no guarantee that all the connections will be patched. This is a potential regression that this change could introduce, if any customers were relying on `dd-trace-py` patching all connections regardless of when `patch` is called.

**Comment by brettlangdon** [comment]
> All the settings.py that I have seen do not import django so shouldn't open any connections to any DB at that point. Of course, if patch is called once the application has been started, there is no guarantee that all the connections will be patched. This is a potential regression that this change could introduce, if any customers were relying on dd-trace-py patching all connections regardless of when patch is called.

That should be fine. I would say patching too late for any integration will likely introduce unintended side effects. As long as the common/accepted approaches (`ddtrace-run`, and calling `patch_all()` inside `setttings.py`) can't cause a connection to not be leaked. Then we are good.

Are there any test cases of value we might be missing to help confirm this?

File: tests/contrib/django/test_django.py [hunk 1]
```diff
@@ -1647,3 +1648,11 @@ def start_response(status, headers):
...
         else:
             assert root.resource == "GET tests.contrib.django.views.error_500"
+
+
+@pytest.mark.django_db
+def test_connections_patched():
+    from django.db import connections
+
+    for conn in connections.all():
+        assert isinstance(conn.cursor, wrapt.ObjectProxy)
```
Comment (brettlangdon) [inline tests/contrib/django/test_django.py:1666]: missing newline at eof here?

(do we care about that?)

File: tests/contrib/django/test_django.py [hunk 2]
```diff
@@ -1647,3 +1648,11 @@ def start_response(status, headers):
             assert root.resource == "GET ^error-500/$"
         else:
             assert root.resource == "GET tests.contrib.django.views.error_500"
+
+
+@pytest.mark.django_db
+def test_connections_patched():
+    from django.db import connections
+
+    for conn in connections.all():
```
Comment (brettlangdon) [inline tests/contrib/django/test_django.py:1665]: we do have connections here right?

e.g. we should be able to do `assert len(connections.all()) == X` or just `assert connections.all()` ?

Comment (P403n1x87) [reply tests/contrib/django/test_django.py:1665]: Ah yeh I checked that there was at least one manually (it's called `default`) but forgot to add the assert! Will add it now.

**Review by brettlangdon — DISMISSED** [review]

**Review by majorgreys — DISMISSED** [review]
👍🏽

**Review by Kyle-Verhoog — COMMENTED** [review]
Just one thought on the release note, otherwise LGTM

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Pull request merged**